File: securehome/smart_home.c
/*
 * Copyright (C) Your copyright notice.
 *
 * Author: Onur Dundar
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <fcntl.h>
#include <string.h>
#include <termios.h>
#include <pthread.h>
#include <sys/ioctl.h>
#include <time.h>
#include "serial.h"
#include "thermometer.h"
#include "message.h"
#include "device.h"

int device_file;
const char* serial_device_path = "/dev/ttyUSB0";

pthread_t reader;
pthread_t device;

void execute_command(int choice);
void user_menu();
uint8_t pick_node();
uint8_t pick_binary_sensor();
uint8_t pick_multilevel_sensor();
uint8_t switch_on_off();

void* reader_thread(void *arg) {
	while (1) {
		uint8_t data[256];
		int m_length = read_from_serial_device(device_file, data);
		if (m_length > 0) {
			handle_incoming_message(device_file, data, m_length);
		}
	}
	return NULL;
}

int main(int argc, char* argv[]) {
	//Open Serial Device
	device_file = open_serial_device(serial_device_path);
	if (device_file < 0) {
		printf("Can't Open Serial Device %s", serial_device_path);
		return EXIT_FAILURE;
	}
	/**
	 * Create Reader Thread
	 */
	int err = pthread_create(&reader, NULL, &reader_thread, NULL);
	if (err != 0) {
		close_serial_device(device_file);
		printf("Can't create Thread\n");
		return -1;
	}
	/**
	 * Create Device Thread
	 */
	err = pthread_create(&reader, NULL, &update_status, device_file);
	if (err != 0) {
		close_serial_device(device_file);
		printf("Can't create Thread\n");
		return -1;
	}
	printf("Starting Home Manager\n");
	int choice = 0;
	user_menu();
	while (choice != 14) {
		printf("Enter Command to Execute:");
		scanf("%d", &choice);
		execute_command(choice);
	}

	if (pthread_join(reader, NULL)) {
		fprintf(stderr, "Error joining thread\n");
		return 2;
	}
	close_serial_device(device_file);

	return EXIT_SUCCESS;
}

void execute_command(int choice) {
	uint8_t nodeid;
	uint8_t sensor;
	switch (choice) {
	case 1:
		printf("Current Temperature is: %f Celcius\n", read_temperature());
		break;
	case 2:
		printf("Relative Humidity is  : %f percent \n", read_humidity());
		break;
	case 3:
		nodeid = pick_node();
		get_device_status(device_file, nodeid, 3);
		break;
	case 4:
		nodeid = pick_node();
		get_binary_switch_status(device_file, nodeid, 4);
		break;
	case 5:
		nodeid = pick_node();
		sensor = pick_binary_sensor();
		get_binary_sensor_value(device_file, nodeid, sensor, 5);
		break;
	case 6:
		nodeid = pick_node();
		sensor = pick_multilevel_sensor();
		get_multilevel_sensor_value(device_file, nodeid, sensor, 6);
		break;
	case 7:
		nodeid = pick_node();
		get_node_power_level(device_file, nodeid, 7);
		break;
	case 8:
		nodeid = pick_node();
		get_battery_level(device_file, nodeid, 8);
		break;
	case 9:
		get_meter_level(device_file, WallPlugNodeID, POWER, 9);
		break;
	case 10:
		get_meter_level(device_file, WallPlugNodeID, ENERGY, 10);
		break;
	case 11:
		nodeid = pick_node();
		uint8_t on_off = switch_on_off();
		binary_switch_on_off(device_file, nodeid, on_off, 10);
		break;
	case 12:
		printf("Quitting.....");
		break;
	default:
		break;
	}
}

void user_menu() {
	printf("1  : Show Current Temperature						\n");
	printf("2  : Show Current Relative Humidity					\n");
	printf("3  : Get Current Status Node						\n");
	printf("4  : Get Current Switch Status on Node				\n");
	printf("5  : Get Current Sensor Status on Node			  	\n");
	printf("6  : Get Current Multilevel Sensor Status on Node 	\n");
	printf("7  : Get Power Level of Node						\n");
	printf("8  : Get Device Battery Level						\n");
	printf("9  : Get Current Power Consumption					\n");
	printf("10 : Get Energy Consumption 						\n");
	printf("11 : Switch On-Off Node							\n");
	printf("12 : Quit											\n");
}

uint8_t pick_node() {
	int node;
	printf("Multi-Sensor Node   :1\n");
	printf("Wall-Plug           :2\n");
	printf("Lamp Holder         :3\n");
	printf("Flood Detector      :4\n");
	printf("Select Device : ");
	scanf("%d", &node);
	switch (node) {
	case 1:
		node = MultiSensorNodeID;
		break;
	case 2:
		node = WallPlugNodeID;
		break;
	case 3:
		node = LampHolderNodeID;
		break;
	case 4:
		node = FloodSensorNodeID;
		break;
	default:
		printf("invalid node\n");
		node = 1;
		break;
	}
	return (uint8_t) node;
}
uint8_t pick_binary_sensor() {
	int sensor;
	printf("Tamper              :1\n");
	printf("Door/Window         :2\n");
	printf("Motion              :3\n");
	printf("Select Binary Sensor: ");
	scanf("%d", &sensor);
	switch (sensor) {
	case 1:
		sensor = TAMPER_SENSOR;
		break;
	case 2:
		sensor = DOOR_WINDOW_SENSOR;
		break;
	case 3:
		sensor = MOTION_DETECTION_SENSOR;
		break;
	default:
		printf("invalid sensor\n");
		sensor = 1;
		break;
	}
	return (uint8_t) sensor;
}
uint8_t pick_multilevel_sensor() {
	int sensor;
	printf("Temperature         :1\n");
	printf("Illumination        :2\n");
	printf("Power               :3\n");
	printf("Select MultiLevel Sensor: ");
	scanf("%d", &sensor);
	switch (sensor) {
	case 1:
		sensor = TEMPERATURE_SENSOR;
		break;
	case 2:
		sensor = LUMINANCE_SENSOR;
		break;
	case 3:
		sensor = POWER_SENSOR;
		break;
	default:
		printf("invalid sensor\n");
		sensor = 1;
		break;
	}
	return (uint8_t) sensor;
}
uint8_t switch_on_off() {
	int res;
	printf("Switch ON   : 1\n");
	printf("Switch OFF  : 0\n");
	printf("Select: ");
	scanf("%d", &res);
	if (res == 1) {
		res = ON;
	} else if (res == 0) {
		res = OFF;
	} else {
		printf("invalid\n");
		res = -1;
	}
	return (uint8_t) res;
}

File: smarthome_app/src/com/galileha/smarthome/SmartHome.java
/*
 * Copyright (C) Your copyright notice.
 *
 * Author: Onur Dundar
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.galileha.smarthome;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.MalformedURLException;
import java.net.Socket;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Scanner;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

/**
 * Activity Class to Populate Views for Android Application
 * Create threads for network operations
 */
public class SmartHome extends Activity {

	private URL homeJSONUrl;
	private URL commandsJSONUrl;
	private String lastUpdateDate;
	private JSONObject homeInfo;
	private String[] commands;
	private String[] deviceNames;
	private String[] deviceStatus;
	private Integer[] deviceImages = { R.drawable.sht11, R.drawable.philio,
			R.drawable.fibaro, R.drawable.everspring, R.drawable.flood,
			R.drawable.mq9, R.drawable.dlink };

	private Socket galileSocket;
	private static final int GALILEOPORT = 3500;
	private static final String GALILEO_IP = "192.168.2.235";

	volatile public boolean filesdownloaded = false;
	volatile public boolean socketcreated = false;

	ArrayList<DeviceItem> listItems;

	ListView deviceListView;
	TextView updateDateText;
	
	/**
	 * Create Socket Connection Thread
	 */
	Thread socketThread = new Thread(new Runnable() {
		@Override
		public void run() {
			// TODO Auto-generated method stub
			InetAddress serverAddr;
			try {
				serverAddr = InetAddress.getByName(GALILEO_IP);
				galileSocket = new Socket(serverAddr, GALILEOPORT);
				socketcreated = true;
			} catch (UnknownHostException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (IOException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
	});
	/**
	 * JSON Worker to Fetch JSON Files from Intel Galileo
	 */
	Thread jsonFetcher = new Thread(new Runnable() {
		@Override
		public void run() {
			// TODO Auto-generated method stub
			try {
				// Initialize URLs
				homeJSONUrl = new URL("http://192.168.2.235:3000/home.json");
				commandsJSONUrl = new URL(
						"http://192.168.2.235:3000/commands.json");
				getHomeStatus();
				getCommandsList();
				// Fill String Arrays
				initalizeHomeArray();
				fillListAdapter();
				filesdownloaded = true;
				Log.d("JSON", "JSON Files Fetched....");
			} catch (MalformedURLException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (IOException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (JSONException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
	});
	
	/**
	 * Update Adapter
	 * Periodically Read the home.json File and Update ListView Adapter
	 */
	Thread updater = new Thread(new Runnable() {

		@Override
		public void run() {
			// TODO Auto-generated method stub
			while (true) {
				try {
					try {
						Thread.sleep(10000);
					} catch (InterruptedException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
					// Initialize URLs
					homeJSONUrl = new URL("http://192.168.2.235:3000/home.json");
					getHomeStatus();
					// Fill String Arrays
					updateHomeArray();
					// Clear Adapter
					listItems.clear();
					fillListAdapter();
					Log.d("JSON", "Updated ListView Adapter....");
				} catch (MalformedURLException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				} catch (IOException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				} catch (JSONException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		}
	});

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_smart_home);
		deviceListView = (ListView) findViewById(R.id.list);
		updateDateText = (TextView) findViewById(R.id.updateDate);
		listItems = new ArrayList<DeviceItem>();
		jsonFetcher.start();
		while (!filesdownloaded)
			;
		HomeDeviceAdapter deviceAdapter = new HomeDeviceAdapter(this, listItems);
		deviceListView.setAdapter(deviceAdapter);
		deviceListView.setOnItemClickListener(new OnItemClickListener() {
			@Override
			public void onItemClick(AdapterView<?> parent, View view,
					int position, long id) {
				JSONObject homeDevices;
				JSONArray devices;
				JSONObject clickedObject = null;
				try {
					homeDevices = homeInfo.getJSONObject("home");
					devices = (JSONArray) homeDevices.getJSONArray("device");
					clickedObject = devices.getJSONObject(position);
				} catch (JSONException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
				// Switch ON/OFF Wall Plug or LampHolder
				if (position == 2) {
					try {
						if (clickedObject.getString("switch").compareTo("ON") == 0) {
							writeToSocket(commands[5]);
						} else {
							writeToSocket(commands[4]);
						}
					} catch (JSONException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
				} else if (position == 3) {
					try {
						if (clickedObject.getString("switch").compareTo("ON") == 0) {
							writeToSocket(commands[7]);
						} else {
							writeToSocket(commands[6]);
						}
					} catch (JSONException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
				} else {
					Toast.makeText(SmartHome.this,
							"No Available Command for Selected Device",
							Toast.LENGTH_LONG).show();
				}
			}
		});
		updateDateText.setText(lastUpdateDate);
		socketThread.start();
		while (!socketcreated)
			;
		updater.start();
		Log.d("SOCKET", "Socket Thread Started");
	}

	/**
	 * Read home.json from Server
	 * 
	 * @param URL
	 * @return
	 * @throws IOException
	 *             , MalformedURLException, JSONException
	 */
	public void getHomeStatus() throws IOException, MalformedURLException,
			JSONException {
		// Set URL
		// Connect to Intel Galileo get Device Status
		HttpURLConnection httpCon = (HttpURLConnection) homeJSONUrl
				.openConnection();
		httpCon.setReadTimeout(10000);
		httpCon.setConnectTimeout(15000);
		httpCon.setRequestMethod("GET");
		httpCon.setDoInput(true);
		httpCon.connect();
		// Read JSON File as InputStream
		InputStream readStream = httpCon.getInputStream();
		Scanner scan = new Scanner(readStream).useDelimiter("\\A");
		// Set stream to String
		String jsonFile = scan.hasNext() ? scan.next() : "";
		// Initialize serveFile as read string
		homeInfo = new JSONObject(jsonFile);
		httpCon.disconnect();
	}

	/**
	 * Initialize String Array from JSON
	 * 
	 * @throws JSONException
	 */
	public void initalizeHomeArray() throws JSONException {
		JSONObject homeDevices = homeInfo.getJSONObject("home");
		lastUpdateDate = homeDevices.getString("last_update_date");
		JSONArray devArray = (JSONArray) homeDevices.getJSONArray("device");
		int numOfDevices = devArray.length();
		deviceNames = new String[numOfDevices];
		deviceStatus = new String[numOfDevices];
		// Fill the Name Array
		for (int i = 0; i < numOfDevices; i++) {
			JSONObject devObject = (JSONObject) devArray.get(i);
			deviceNames[i] = devObject.getString("name");
		}
		// Fill the Status Array
		updateHomeArray();
	}

	/**
	 * Update Home Status Array with Reading from JSON Object
	 * 
	 * @throws JSONException
	 */
	public void updateHomeArray() throws JSONException {
		JSONObject homeDevices = homeInfo.getJSONObject("home");
		lastUpdateDate = homeDevices.getString("last_update_date");
		JSONArray devices = (JSONArray) homeDevices.getJSONArray("device");
		// SHT11
		JSONObject devObject = (JSONObject) devices.get(0);
		deviceStatus[0] = new String("Status: " + devObject.getString("status")
				+ "\nTemp: " + devObject.getString("temperature")
				+ "\nHumidity: " + devObject.getString("humidity"));
		Log.d("HOME", deviceStatus[0]);
		// Philio
		devObject = (JSONObject) devices.get(1);
		deviceStatus[1] = new String("Status: " + devObject.getString("status")
				+ "\nTemp: " + devObject.getString("temperature")
				+ "\nLumination: " + devObject.getString("lumination")
				+ "\nMotion: " + devObject.getString("motion")
				+ "\nDoor/Window: " + devObject.getString("door")
				+ "\nBattery: " + devObject.getString("battery")
				+ "\nPowerLevel: " + devObject.getString("power_level"));
		Log.d("HOME", deviceStatus[1]);
		// Wall Plug
		devObject = (JSONObject) devices.get(2);
		deviceStatus[2] = new String("Status: " + devObject.getString("status")
				+ "\nSwitch: " + devObject.getString("switch")
				+ "\nCurrent Consumption: "
				+ devObject.getString("power_meter")
				+ "\nCumulative Consumption: "
				+ devObject.getString("energy_meter") + "\nPowerLevel: "
				+ devObject.getString("power_level"));
		// Lamp Holder
		devObject = (JSONObject) devices.get(3);
		deviceStatus[3] = new String("Status: " + devObject.getString("status")
				+ "\nSwitch: " + devObject.getString("switch")
				+ "\nPowerLevel: " + devObject.getString("power_level"));
		// Flood Detector
		devObject = (JSONObject) devices.get(4);
		deviceStatus[4] = new String("Status: " + devObject.getString("status")
				+ "\nFlood: " + devObject.getString("flood") + "\nBattery: "
				+ devObject.getString("battery") + "\nPowerLevel: "
				+ devObject.getString("power_level"));
		// GAS Detector
		devObject = (JSONObject) devices.get(5);
		deviceStatus[5] = new String("Status: " + devObject.getString("status")
				+ "\nGas Level: " + devObject.getString("gas_level"));
		// Network Camera
		devObject = (JSONObject) devices.get(6);
		deviceStatus[6] = new String("Status: " + devObject.getString("status")
				+ "\nPort: " + devObject.getString("port"));
	}

	/**
	 * Get Command List JSON File From Server
	 * 
	 * @throws IOException
	 * @throws MalformedURLException
	 * @throws JSONException
	 */
	public void getCommandsList() throws IOException, MalformedURLException,
			JSONException {
		// Connect to Intel Galileo get Commands List
		HttpURLConnection httpCon = (HttpURLConnection) commandsJSONUrl
				.openConnection();
		httpCon.setReadTimeout(10000);
		httpCon.setConnectTimeout(15000);
		httpCon.setRequestMethod("GET");
		httpCon.setDoInput(true);
		httpCon.connect();
		// Read JSON File as InputStream
		InputStream readCommand = httpCon.getInputStream();
		Scanner scanCommand = new Scanner(readCommand).useDelimiter("\\A");
		// Set stream to String
		String commandFile = scanCommand.hasNext() ? scanCommand.next() : "";
		// Initialize serveFile as read string
		JSONObject commandsList = new JSONObject(commandFile);
		JSONObject temp = (JSONObject) commandsList.get("commands");
		JSONArray comArray = (JSONArray) temp.getJSONArray("command");
		int numberOfCommands = comArray.length();
		commands = new String[numberOfCommands];
		// Fill the Array
		for (int i = 0; i < numberOfCommands; i++) {
			JSONObject commandObject = (JSONObject) comArray.get(i);
			commands[i] = commandObject.getString("text");
		}
		Log.d("JSON", "Loaded " + commands[2]);
		httpCon.disconnect();
	}

	/**
	 * Fill the ListView Adapter
	 */
	public void fillListAdapter() {
		int numberOfItems = deviceStatus.length;
		for (int i = 0; i < numberOfItems; i++) {
			DeviceItem newItem = new DeviceItem();
			newItem.setDeviceName(deviceNames[i]);
			newItem.setDeviceStatus(deviceStatus[i]);
			newItem.setDeviceImage(deviceImages[i]);
			listItems.add(newItem);
		}
	}

	/**
	 * Write Command to Intel Galileo Socket
	 * 
	 * @param message
	 */
	public void writeToSocket(String message) {
		try {
			PrintWriter out = new PrintWriter(new BufferedWriter(
					new OutputStreamWriter(galileSocket.getOutputStream())),
					true);
			out.println(message);
			Log.d("SOCKET", "Message " + message + " Sent to Socket");
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	/**
	 * Update View Items
	 * 
	 * @param v
	 */
	public void onUpdate(View v) {
		updateDateText.invalidate();
		deviceListView.invalidateViews();
		deviceListView.refreshDrawableState();
		updateDateText.setText(lastUpdateDate);
		updateDateText.refreshDrawableState();
		Log.d("JSON", "Updated Views....");
	}
}
